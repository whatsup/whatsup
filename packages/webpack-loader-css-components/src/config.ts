import { cosmiconfig } from 'cosmiconfig'
import { ProcessOptions, PluginCreator, Plugin } from 'postcss'
import { LoaderContext } from './index'

type PluginsConfig = { [k: string]: boolean | PluginOptions } | (string | [string, PluginOptions])[]

type PluginOptions = {
    [k: string]: any
}

type Config = {
    plugins?: PluginsConfig
    processOptions?: ProcessOptions
}

const loadRawConfig = async (loaderContext: LoaderContext) => {
    const explorer = cosmiconfig('postcss')

    try {
        const found = await explorer.search()

        if (found === null || found.isEmpty) {
            throw 'Config is empty'
        }

        const { config, filepath } = found

        loaderContext.addBuildDependency(filepath)
        loaderContext.addDependency(filepath)

        if (typeof config === 'function') {
            const ctx = {
                env: loaderContext.mode,
                mode: loaderContext.mode,
                file: loaderContext.resourcePath,
                webpackLoaderContext: loaderContext,
                options: {},
            }

            return config(ctx) as Config
        }

        return config as Config
    } catch (e) {
        return {} as Config
    }
}

function* readPlugins(plugins: PluginsConfig): Generator<[string, PluginOptions], void, unknown> {
    if (Array.isArray(plugins)) {
        for (const item of plugins) {
            if (Array.isArray(item)) {
                const [plugin, options = {}] = item

                yield [plugin, options]
            } else {
                const plugin = item
                const options = {} as PluginOptions

                yield [plugin, options]
            }
        }
    } else {
        for (const [plugin, val] of Object.entries(plugins)) {
            if (!!val) {
                const options = typeof val === 'object' ? val : {}

                yield [plugin, options]
            }
        }
    }
}

const ignorePlugins = ['postcss-import', 'postcss-url', 'postcss-modules']

const preparePlugins = (plugins: PluginsConfig) => {
    const result = [] as Plugin[]

    for (const [plugin, options] of readPlugins(plugins)) {
        if (!ignorePlugins.includes(plugin)) {
            const module = require(plugin) as PluginCreator<PluginOptions>

            result.push(module(options) as Plugin)
        }
    }

    return result
}

const prepareProcessOptions = (options: ProcessOptions, loaderContext: LoaderContext) => {
    options.from = loaderContext.resourcePath
    options.to = undefined

    return options
}

export const loadConfig = async (loaderContext: LoaderContext) => {
    const raw = await loadRawConfig(loaderContext)
    const plugins = preparePlugins(raw.plugins || [])
    const processOptions = prepareProcessOptions(raw.processOptions || {}, loaderContext)

    return { plugins, processOptions }
}
