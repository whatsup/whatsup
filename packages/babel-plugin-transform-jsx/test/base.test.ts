import { transform } from './config/transform'

describe('transform types', function () {
    it('Should transform div', function () {
        expect(transform('<div>foobar</div>')).toMatchSnapshot()
    })

    it('Should transform User', function () {
        expect(transform('<User>foobar</User>')).toMatchSnapshot()
    })

    it('Should transform svg', function () {
        expect(transform('<svg>foobar</svg>')).toMatchSnapshot()
    })

    it('Should transform fragment', function () {
        expect(transform('<>foobar</>')).toMatchSnapshot()
    })

    it('Should throw error on namespaced name', function () {
        expect(() => transform('<ns:div>foobar</ns:div>')).toThrowError(
            'Namespaced openingElement name is not supported'
        )
    })

    it('Should throw error when name is member expression', function () {
        expect(() => transform('<ns.div>foobar</ns.div>')).toThrowError(
            'OpeningElement name as member expression is not supported'
        )
    })
})
