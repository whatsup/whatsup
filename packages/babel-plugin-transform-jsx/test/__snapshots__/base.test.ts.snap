// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`transform types Should transform User 1`] = `
"import { component as _component } from \\"@fract/jsx\\";

_component(User, \\"UniqueId\\", undefined, undefined, [\\"foobar\\"]);"
`;

exports[`transform types Should transform div 1`] = `
"import { html as _html } from \\"@fract/jsx\\";

_html(\\"div\\", \\"UniqueId\\", undefined, undefined, [\\"foobar\\"]);"
`;

exports[`transform types Should transform fragment 1`] = `
"import { Fragment as _Fragment } from \\"@fract/jsx\\";
import { component as _component } from \\"@fract/jsx\\";

_component(_Fragment, \\"UniqueId\\", undefined, undefined, [\\"foobar\\"]);"
`;

exports[`transform types Should transform svg 1`] = `
"import { svg as _svg } from \\"@fract/jsx\\";

_svg(\\"svg\\", \\"UniqueId\\", undefined, undefined, [\\"foobar\\"]);"
`;
